text/0006-sampling.md

```diff
@@ -0,0 +1,302 @@
+
+ * They care about a variety of things, including efficiency, cost effectiveness, and being able to
+ gather spans in a way that makes sense for them.
+
+**Solution:**
+ * Infrastructure owners receive information attached to the span, after sampling hooks have already
+ been run.
+
+## Internal details
+In Dapper based systems (or systems without a deferred sampling decision) all exported spans are 
+stored to the backend, thus some of these systems usually don't scale to a high volume of traces,
+or the cost to store all the Spans may be too high. In order to support this use-case and to 
+ensure the quality of the data we send, OpenTelemetry needs to natively support sampling with some
+requirements:
+ * Send as many complete traces as possible. Sending just a subset of the spans from a trace is 
+ less useful because in this case the interaction between the spans may miss.
+ * Allow application operator to configure the sampling frequency.
+ 
+For new modern systems that need to collect all the Spans and later may or may not do a deferred 
+sampling decision, OpenTelemetry needs to natively support a way to configure the library to 
+collect and export all the Spans. This is possible even though OpenTelemetry supports sampling by
+setting a default config to always collect all the spans.
+
+### Sampling flags
+OpenTelemetry API has two flags/properties:
+ * `RecordEvents`
+   * This property is exposed in the `Span` interface (e.g. `Span.isRecordingEvents()`).
+   * If `true` the current `Span` records tracing events (attributes, events, status, etc.), 
+   otherwise all tracing events are dropped.
+   * Users can use this property to determine if expensive trace events can be avoided.
+ * `SampledFlag`
+   * This flag is propagated via the `TraceOptions` to the child Spans (e.g.
+   `TraceOptions.isSampled()`). For more details see the w3c definition [here][trace-flags].
+   * In Dapper based systems this is equivalent to `Span` being `sampled` and exported.
+   
+The flag combination `SampledFlag == false` and `RecordEvents == true` means that the current `Span`
+does record tracing events, but most likely the child `Span` will not. This combination is 
+necessary because:
+ * Allow users to control recording for individual Spans.
+ * OpenCensus has this to support z-pages, so we need to keep backwards compatibility.
+
+The flag combination `SampledFlag == true` and `RecordEvents == false` can cause gaps in the 
+distributed trace, and because of this OpenTelemetry API should NOT allow this combination.
+
+It is safe to assume that users of the API should only access the `RecordEvents` property when 
+instrumenting code and never access `SampledFlag` unless used in context propagators.
+
+### SamplingHint
+This is a new concept added in the OpenTelemetry API that allows to suggest sampling hints to the
+implementation of the API:
+ * `UNSPECIFIED`
+   * This is the default option.
+   * No suggestion is made.
```

Comment:
I assume the hint object is going to be an optional parameter to start-span function. If so, isn't UNSPECIFIED the same as "hint == null" (i.e. no hint from the user)? It seems like an unnecessary cognitive overload to have this constant.

Reply:
I am a dev that does not like NULL as a default. The reason is that your code will look like this:

```
if (hint == null) {
  do something for default
}
switch (hint) {
  case RECORD:
    ...
...
}
```

With non-null value the code is simpler:
```
switch (hint) {
  case UNSPECIFIED {
    do something for default
    break;
  }
  case RECORD:
    ...
...
}
```

It is more readable in my opinion.

Reply:
Whether to keep the NULL or not is an internal implementation choice, and also language specific -- you could use Option, you could catch NULL in `default:` part of the switch, or in Python we don't even have enums so any private value can be used as an UNSPECIFIED sentinel. You could also keep a separate isHintSet boolean. There are lots of options, all having to deal with the possibility of user never invoking `withSamplingHint()`, which I assume we're not suggesting to make a required arg.

But exposing UNSPECIFIED at the API level allows the user to do meaningless thing like
```
spanBuilder.withSamplingHint(SamplingHind.UNSPECIFIED)
```

I would choose clean API over any implementation concerns.

Reply:
I don't have a strong opinion, and definitely all the options you proposed are valid. To clarify: Is your suggestion to remove UNSPECIFIED from the options that user has in the API to set the hint? In other words user has two options: does not set a hint OR sets one of the hints that are meaningful.

Reply:
> In other words user has two options: does not set a hint OR sets one of the hints that are meaningful.

yes